lib/tlsca/ca.go

```diff
@@ -110,37 +128,82 @@ func (id *Identity) Subject() (pkix.Name, error) {
 	subject.Organization = append([]string{}, id.Groups...)
 	subject.OrganizationalUnit = append([]string{}, id.Usage...)
 	subject.Locality = append([]string{}, id.Principals...)
+
+	// DELETE IN (4.4.0)
+	// Groups are marshaled to both ASN1 extension
+	// and old Province section, for backwards-compatibility,
+	// however begin migration to ASN1 extensions in the future
+	// for this and other properties
 	subject.Province = append([]string{}, id.KubernetesGroups...)
 	subject.StreetAddress = []string{id.RouteToCluster}
 	subject.PostalCode = []string{string(rawTraits)}
 
+	for i := range id.KubernetesUsers {
+		kubeUser := id.KubernetesUsers[i]
+		subject.ExtraNames = append(subject.ExtraNames,
+			pkix.AttributeTypeAndValue{
+				Type:  KubeUsersASN1ExtensionOID,
+				Value: kubeUser,
+			})
+	}
+
+	for i := range id.KubernetesGroups {
+		kubeGroup := id.KubernetesGroups[i]
+		subject.ExtraNames = append(subject.ExtraNames,
+			pkix.AttributeTypeAndValue{
+				Type:  KubeGroupsASN1ExtensionOID,
+				Value: kubeGroup,
+			})
+	}
+
 	return subject, nil
 }
 
 // FromSubject returns identity from subject name
 func FromSubject(subject pkix.Name, expires time.Time) (*Identity, error) {
-	i := &Identity{
-		Username:         subject.CommonName,
-		Groups:           subject.Organization,
-		Usage:            subject.OrganizationalUnit,
-		Principals:       subject.Locality,
-		KubernetesGroups: subject.Province,
-		Expires:          expires,
+	id := &Identity{
+		Username:   subject.CommonName,
+		Groups:     subject.Organization,
+		Usage:      subject.OrganizationalUnit,
+		Principals: subject.Locality,
+		Expires:    expires,
 	}
 	if len(subject.StreetAddress) > 0 {
-		i.RouteToCluster = subject.StreetAddress[0]
+		id.RouteToCluster = subject.StreetAddress[0]
 	}
 	if len(subject.PostalCode) > 0 {
-		err := wrappers.UnmarshalTraits([]byte(subject.PostalCode[0]), &i.Traits)
+		err := wrappers.UnmarshalTraits([]byte(subject.PostalCode[0]), &id.Traits)
 		if err != nil {
 			return nil, trace.Wrap(err)
 		}
 	}
 
-	if err := i.CheckAndSetDefaults(); err != nil {
+	for _, attr := range subject.Names {
+		switch {
+		case attr.Type.Equal(KubeUsersASN1ExtensionOID):
+			val, ok := attr.Value.(string)
+			if ok {
+				id.KubernetesUsers = append(id.KubernetesUsers, val)
+			}
+		case attr.Type.Equal(KubeGroupsASN1ExtensionOID):
+			val, ok := attr.Value.(string)
+			if ok {
+				id.KubernetesGroups = append(id.KubernetesGroups, val)
+			}
+		}
+	}
+
+	// DELETE IN(4.5.0): This logic is using Province field
+	// from subject in case if Kubernetes groups were not populated
+	// from ASN1 extension
+	if len(id.KubernetesGroups) == 0 {
+		id.KubernetesGroups = subject.Province
+	}
```

Comment:
We should either keep this around longer or make a non-empty `Province` become an error.  Silently pruning kubernetes groups on older certs (such as those exported with `tctl auth sign`) would be very confusing.

Reply:
I'm ok to keep this longer let's say to 5.0 release